### PR TITLE
Enable nullable on server transports

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/dotnet/aspnetcore/issues/5680

### CODE REVIEW

File: src/Servers/Connections.Abstractions/src/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -102,7 +102,7 @@ Microsoft.AspNetCore.Connections.IConnectionBuilder.Use(System.Func<Microsoft.As
 Microsoft.AspNetCore.Connections.IConnectionFactory
 Microsoft.AspNetCore.Connections.IConnectionFactory.ConnectAsync(System.Net.EndPoint! endpoint, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.ConnectionContext!>
 Microsoft.AspNetCore.Connections.IConnectionListener
-Microsoft.AspNetCore.Connections.IConnectionListener.AcceptAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.ConnectionContext!>
+Microsoft.AspNetCore.Connections.IConnectionListener.AcceptAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.ConnectionContext?>
```
Comment (JamesNK) [inline src/Servers/Connections.Abstractions/src/PublicAPI.Shipped.txt:105]: Nullable "breaking change" @pranavkm

File: src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -6,7 +6,7 @@ Microsoft.AspNetCore.Connections.Experimental.IMultiplexedConnectionBuilder.Use(
...
 Microsoft.AspNetCore.Connections.Experimental.IMultiplexedConnectionListener
-Microsoft.AspNetCore.Connections.Experimental.IMultiplexedConnectionListener.AcceptAsync(Microsoft.AspNetCore.Http.Features.IFeatureCollection? features = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionContext!>
+Microsoft.AspNetCore.Connections.Experimental.IMultiplexedConnectionListener.AcceptAsync(Microsoft.AspNetCore.Http.Features.IFeatureCollection? features = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionContext?>
```
Comment (JamesNK) [inline src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:9]: Nullable "breaking change" (although the package is experimental) @pranavkm

File: src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -19,5 +19,5 @@ Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionBuilder.Use(S
 Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionContext
 Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionContext.MultiplexedConnectionContext() -> void
 Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionDelegate
-abstract Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionContext.AcceptAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.ConnectionContext!>
+abstract Microsoft.AspNetCore.Connections.Experimental.MultiplexedConnectionContext.AcceptAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.ConnectionContext?>
```
Comment (JamesNK) [inline src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:22]: Nullable "breaking change" (although the package is experimental) @pranavkm

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs [hunk 4]
```diff
@@ -26,6 +26,11 @@ internal class QuicConnectionListener : IMultiplexedConnectionListener, IAsyncDi
 
         public QuicConnectionListener(QuicTransportOptions options, IQuicTrace log, EndPoint endpoint)
         {
+            if (options.Alpn == null)
+            {
+                throw new InvalidOperationException("QuicTransportOptions.Alpn must be configured with a value.");
```
Comment (JamesNK) [inline src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:31]: What should happen if no ALPN is configured?

File: src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt (hunk 3 above)
Comment (halter73) [reply src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:22]: This [wasn't public in .NET 5](https://github.com/dotnet/aspnetcore/pull/27506), so we should be completely fine here.

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs (hunk 4 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:31]: I think this is fine, I'm not sure what we can put as a default.

Comment (halter73) [reply src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:31]: @jkotalik 

I think this is good enough for now. The whole experience needs to be improved:

https://github.com/dotnet/aspnetcore/blob/9cfb13398c3296f51114a4b9eb01d69b93b5195c/src/Servers/Kestrel/samples/Http3SampleApp/Program.cs#L28-L33

It also looks like the IdleTimeout defaults to zero which probably isn't ideal.

File: src/Servers/Kestrel/Transport.Quic/src/QuicTransportOptions.cs [hunk 5]
```diff
@@ -25,22 +25,22 @@ public class QuicTransportOptions
         /// <summary>
         /// The Application Layer Protocol Negotiation string.
         /// </summary>
-        public string Alpn { get; set; }
+        public string? Alpn { get; set; }
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Quic/src/QuicTransportOptions.cs:28]: I'm guessing that if we ship this it won't be nullable. But given there's no default value at the moment, I think this is fine for now.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs [hunk 6]
```diff
@@ -102,19 +103,22 @@ void BindSocket()
...
         }
 
-        public async ValueTask<ConnectionContext> AcceptAsync(CancellationToken cancellationToken = default)
+        public async ValueTask<ConnectionContext?> AcceptAsync(CancellationToken cancellationToken = default)
         {
             while (true)
             {
                 try
                 {
+                    Debug.Assert(_listenSocket != null, "BindAsync must have been called.");
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:120]: Nit:

```suggestion
                    Debug.Assert(_listenSocket != null, "Bind must be called first.");
```

File: src/Servers/Connections.Abstractions/src/PublicAPI.Shipped.txt (hunk 1 above)
Comment (pranavkm) [reply src/Servers/Connections.Abstractions/src/PublicAPI.Shipped.txt:105]: Could you add a comment to https://github.com/aspnet/Announcements/issues/444?

File: src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (pranavkm) [reply src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:9]: Changes in Unshipped would be non-breaking.

**Review by halter73 — APPROVED** [review]

**Pull request merged**